edfi-repo-auditor/edfi_repo_auditor/auditor.py

```diff
@@ -0,0 +1,163 @@
+
+
+def audit_actions(client: GitHubClient, organization: str, repository: str) -> dict:
+    audit_results = {
+        "Has Actions": False,
+        "Uses CodeQL": False,
+        "Uses Allowed list": False,
+        "Uses Test Reporter": False,
+        "Has Unit Tests": False,
+        "Has Linter": False
+    }
+
+    actions = client.get_actions(organization, repository)
+
+    logger.debug(f"Got {actions['total_count']} workflow files")
+
+    audit_results["Has Actions"] = actions["total_count"] > 0
+
+    workflow_paths = [actions["workflows"]["path"] for actions["workflows"] in actions["workflows"]]
+    for file_path in workflow_paths:
+        file_content = client.get_file_content(organization, repository, file_path)
+        if not file_content:
+            logger.debug("File not found")
+            continue
+
+        if not audit_results["Uses CodeQL"]:
+            audit_results["Uses CodeQL"] = "uses: github/codeql-action/analyze" in file_content
+
+        if not audit_results["Uses Allowed list"]:
+            audit_results["Uses Allowed list"] = "uses: ed-fi-alliance-oss/ed-fi-actions/.github/workflows/repository-scanner.yml" in file_content
+
+        if not audit_results["Uses Test Reporter"]:
+            audit_results["Uses Test Reporter"] = "uses: dorny/test-reporter" in file_content
+
+        if not audit_results["Has Unit Tests"]:
+            pattern = re.compile(r"unit.{0,2}test(s)?", flags=re.IGNORECASE)
+            audit_results["Has Unit Tests"] = True if pattern.search(file_content) else False
+
+        if not audit_results["Has Linter"]:
+            pattern = re.compile(r"lint(er)?(s)?(ing)?", flags=re.IGNORECASE)
+            audit_results["Has Linter"] = True if pattern.search(file_content) else False
+
+    return audit_results
+
+
+def get_repo_information(client: GitHubClient, organization: str, repository: str) -> dict:
+    information = client.get_repository_information(organization, repository)
+
+    # Currently,this is only checking if there are alerts, which does not differentiates if dependabot is enabled or not
+    vulnerabilities = [alerts for alerts in information["vulnerabilityAlerts"]["nodes"]
+                       if (alerts["createdAt"] < (datetime.now() - timedelta(ALERTS_WEEKS_SINCE_CREATED * 7)).isoformat() and
+                       alerts["securityVulnerability"]["advisory"]["severity"] in ALERTS_INCLUDED_SEVERITIES)]
+
+    rulesForMain = [rules for rules in information["branchProtectionRules"]["nodes"] if rules["pattern"] == "main"]
+    rules = rulesForMain[0] if rulesForMain else None
+
+    logger.debug(f"Repository information: {information}")
+
+    return {
+        "Requires Signed commits": rules["requiresCommitSignatures"] if rules else False,
+        "Requires Code review": rules["requiresApprovingReviews"] if rules else False,
+        "Requires PR": (rules["requiresApprovingReviews"] and rules["isAdminEnforced"]) if rules else False,
```

Comment:
I'm not sure if I agree about `isAdminEnforced` being required. It was deliberately not included in my requirements. Of course there _is_ merit to using it, so let's think about this for a few minutes/hours before making a hasty decision.

Reply:
OK, let's do this: separate into two rules so that we have a clearer understanding.

* Requires PR
* Admin cannot bypass PR